### PR TITLE
refactor(redis): improve init logic for JedisPooled client

### PR DESCRIPTION
**Pull request by vegetablest**
The Redis AUTH command can use it in two forms. [redis-security-acl](https://redis.io/docs/latest/operate/oss_and_stack/management/security/acl/):

After redis 6.0:
```bash
AUTH <username> <password>
```

Before redis 6.0:
```bash
AUTH <password>
```

So I think it is more reasonable to judge whether `password` is null when initializing the redis client.

### CODE REVIEW

**Comment by langchain4j** [comment]
@vegetablest please elaborate, the description is not clear. Tests are also missing

**Comment by vegetablest** [comment]
> @vegetablest please elaborate, the description is not clear. Tests are also missing

I added two new unit tests. Before the refactoring, executing `testAuthPass` failed.

**Comment by zambrinf** [comment]
@vegetablest can you change that for `RedisChatMemoryStore` too? In that case you can remove the ensureNotBlank for the user

**Comment by vegetablest** [comment]
> @vegetablest can you change that for `RedisChatMemoryStore` too? In that case you can remove the ensureNotBlank for the user

Sorry, I'm too far behind the main branch. I will rebase and then apply the changes to RedisChatMemoryStore shortly.

**Comment by langchain4j** [comment]
Hi @vegetablest, thanks!
 
`constructor_password_null` fails for me:

```
java.lang.AssertionError: 
Expecting code to raise a throwable.

	at dev.langchain4j.store.memory.chat.redis.RedisChatMemoryStoreIT.constructor_password_null(RedisChatMemoryStoreIT.java:203)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

`constructor_user_empty` as well:

```
2024-08-06 12:07:55 [main] redis.clients.jedis.ConnectionFactory.makeObject()
DEBUG: Error while makeObject: redis.clients.jedis.exceptions.JedisAccessControlException: WRONGPASS invalid username-password pair or user is disabled.
	at redis.clients.jedis.Protocol.processError(Protocol.java:101)
	at redis.clients.jedis.Protocol.process(Protocol.java:162)
	at redis.clients.jedis.Protocol.read(Protocol.java:221)
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:351)
	at redis.clients.jedis.Connection.getStatusCodeReply(Connection.java:269)
	at redis.clients.jedis.Connection.helloOrAuth(Connection.java:484)
	at redis.clients.jedis.Connection.initializeFromClientConfig(Connection.java:402)
	at redis.clients.jedis.Connection.<init>(Connection.java:68)
	at redis.clients.jedis.ConnectionFactory.makeObject(ConnectionFactory.java:59)
	at org.apache.commons.pool2.impl.GenericObjectPool.create(GenericObjectPool.java:571)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:298)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:223)
	at redis.clients.jedis.util.Pool.getResource(Pool.java:38)
	at redis.clients.jedis.ConnectionPool.getResource(ConnectionPool.java:29)
	at redis.clients.jedis.ConnectionPool.getResource(ConnectionPool.java:7)
	at redis.clients.jedis.providers.PooledConnectionProvider.getConnection(PooledConnectionProvider.java:63)
	at redis.clients.jedis.UnifiedJedis.<init>(UnifiedJedis.java:98)
	at redis.clients.jedis.JedisPooled.<init>(JedisPooled.java:393)
	at redis.clients.jedis.JedisPooled.<init>(JedisPooled.java:76)
	at redis.clients.jedis.JedisPooled.<init>(JedisPooled.java:72)
	at dev.langchain4j.store.memory.chat.redis.RedisChatMemoryStore.<init>(RedisChatMemoryStore.java:28)
	at dev.langchain4j.store.memory.chat.redis.RedisChatMemoryStore$RedisChatMemoryStoreBuilder.build(RedisChatMemoryStore.java:19)
	at dev.langchain4j.store.memory.chat.redis.RedisChatMemoryStoreIT.lambda$constructor_user_empty$11(RedisChatMemoryStoreIT.java:184)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:892)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1366)
	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1210)
	at dev.langchain4j.store.memory.chat.redis.RedisChatMemoryStoreIT.constructor_user_empty(RedisChatMemoryStoreIT.java:179)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:728)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:218)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:214)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:139)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:198)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:169)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:93)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:58)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:141)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:57)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:103)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:85)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:63)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)

java.lang.AssertionError: 
Expecting code to raise a throwable.

	at dev.langchain4j.store.memory.chat.redis.RedisChatMemoryStoreIT.constructor_user_empty(RedisChatMemoryStoreIT.java:179)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)


```

**Comment by langchain4j** [comment]
@zambrinf could you help with review please?

**Comment by vegetablest** [comment]
I think these two unit tests can be deleted. @zambrinf WDYT?

**Comment by zambrinf** [comment]
@vegetablest yes, but add one test to assure it works with user null and password not null

**Comment by vegetablest** [comment]
> @vegetablest yes, but add one test to assure it works with user null and password not null

I've added two unit tests. Could you please review them when you have time?

File: langchain4j-redis/src/test/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStoreIT.java [hunk 1]
```diff
@@ -56,4 +63,29 @@ protected EmbeddingStore<TextSegment> embeddingStore() {
...
+                .port(redis.getFirstMappedPort())
+                .password("redis-stack")
+                .indexName(randomUUID())
+                .dimension(384)
+                .metadataKeys(createMetadata().toMap().keySet())
+                .build());
+    }
+
+    @Test
+    public void testUnauthorized() {
```
Comment (zambrinf) [inline langchain4j-redis/src/test/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStoreIT.java:77]: you could add this test to the MemoryStore too. And add the wrong password test to the EmbeddingStore.

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public RedisEmbeddingStore(String host,
         ensureNotNull(port, "port");
         ensureNotNull(dimension, "dimension");
 
-        this.client = user == null ? new JedisPooled(host, port) : new JedisPooled(host, port, user, password);
+        this.client = Objects.isNull(password) ? new JedisPooled(host, port) : new JedisPooled(host, port, user, password);
```
Comment (zambrinf) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: maybe use the same logic that is present in `RedisChatMemoryStore`

**Comment by vegetablest** [comment]
@langchain4j @zambrinf Any other suggestions?

**Review by zambrinf — APPROVED** [review]
looks good to me
